TemplePlus/critter.cpp

```diff
@@ -710,6 +710,101 @@ StandPoint LegacyCritterSystem::GetStandPoint(objHndl critter, StandPointType ty
 	return result;
 }
 
+
+int LegacyCritterSystem::GetWaypointsCount(objHndl critter)
+{
+	return (int)objects.getArrayFieldInt64(critter, obj_f_npc_waypoints_idx, 0);
+}
+
+bool LegacyCritterSystem::GetWaypoint(objHndl critter, int index, Waypoint& wp)
+{
+	if (index >= GetWaypointsCount(critter))
+		return false;
+	union conv32
+	{
+		uint32_t u32; // here_write_bits
+		float    f32; // here_read_float
+	} u;
+
+	auto arrSize = objects.getArrayFieldInt64Size(critter, obj_f::obj_f_npc_waypoints_idx);
+	if (arrSize <= 2 + index * 8 + 7)
```

Comment:
Please no magic numbers

Reply:
Waypoint structure is too big to fit 64 bit value, so it is split into array of int64 and saved like this. But first two positions are reserved for length, thus shift 2. Array of Waypoints are converted to array of int64, that's "index".
= length_offset + waypoint_index*waypoint_size+waypoint_size-1

The same approach is here: int64_t v = objects.getArrayFieldInt64(critter, obj_f::obj_f_npc_waypoints_idx, 2 + index * 8 + row);
and in LegacyCritterSystem::SetWaypoint:
obj->SetInt64(obj_f_npc_waypoints_idx, 2 + index * 8 + row, v);

I do not know how to write it down obviously in the code.

Reply:
Unless I'm missing something, you can define a struct
struct WaypointPacked {
uint32_t flags = 0;	
LocAndOffsets location = LocAndOffsets::null;	
float rotation = 0;	
uint8_t anims[8] = { 0, 0, 0, 0, 0, 0, 0, 0 };
uint32_t delay = 0;	// is this really uint btw? sometimes people use -1 to encode special values ;)
}
And this will be aligned (after the length term of course). You could use a pointer to this struct as a raw int64_t data:

WaypointPacked wp;
int64_t * wpRaw = (int64_t*) &wp;
for (int i=0; i < sizeof(WaypointPacked) / sizeof(int64_t); ++i){
  *wpRaw = getInt64(..)
  wpRaw++;
}
BTW are you sure it starts from index = 2? What's in index = 1?

Reply:
First index is count of Waypoints, LegacyCritterSystem::SetWaypointsCount:
obj->SetInt64(obj_f_npc_waypoints_idx, 0, count);

Reply:
@shartte also did uint in his implementation here: https://github.com/GrognardsFromHell/OpenTemple/blob/master/Core/Systems/Waypoints/WaypointSystem.cs

Reply:
@DudeMcDude, I did as you asked:
```
	auto obj = objSystem->GetObject(critter);
	int64_t* wpRaw = (int64_t*)&wp;
	for (int row = 0; row < sizeof(Waypoint) / sizeof(int64_t); row++)
	{
		obj->SetInt64(obj_f_npc_waypoints_idx, 2 + index * 8 + row, *wpRaw);
		wpRaw++;
	}

```
But it started crashing for some specific anim set. Not sure why. Although code does look much better.

Reply:
@DudeMcDude, would you allow this? Perhaps you could come up with some idea...

Reply:
Did you differentiate Waypoint and WaypointPacked? One of them has padding after the flags field.

Reply:
I missed that all right. Thank you @DudeMcDude. It works. I will test more though.

Reply:
BTW, note that WaypointPacked is assumed to have the same size as Waypoint (0×40), i.e. it needs to have some pad fields to align it.